### PR TITLE
Complete: Testing Practice -Odin Project

### PR DESCRIPTION
**Pull request by imahnama**
## In this Project we:
- Install Jest
- Create function Capitalize, reverseString, calculator, caesarCipher and arrayAnalysis.
- Write tests for the above-mentioned functions
- Add ReadMe file
- Add linters for the project

### CODE REVIEW

File: caesarCipher.test.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import caesarCipher from './caesarCipher';
```
Comment (mosaaleb) [inline caesarCipher.test.js:1]: - [x] When I test the string "If he had anything confidential to say" against your caesarCipher function:

I got a failing test:

```js
// ohk > oHk
Expected: "Pm ol ohk hufaopun jvumpkluaphs av zhf"
Received: "Pm ol oHk hufaopun jvumpkluaphs av zhf"
```
To figure out where the problem coming from, try this test:
```js
expect(caesarCipherInstance.convertStringNumbers('aaaaaaaa', 7))
```

**Review by mosaaleb — CHANGES_REQUESTED** [review]
#### Well done @imahnama @OdongAlican :clap:

Most requirements are handled, yet there is a bug in CaesarCipher function. Check out my comment below.

- **[SUGGESTION]** Let's better organize your modules and tests and put them all inside the `src` folder, with subfolders for both your modules and another one for tests.

- [x] There are some missing test-cases in CaesarCipher:
  - It has a test for wrapping from `z` to `a`.
  - It has a test for keeping the same case.

Please fix the below requirement and submit a new code review.


##### Muhammad Ebeid

**Comment by imahnama** [comment]
> #### Well done @imahnama @OdongAlican 👏
> Most requirements are handled, yet there is a bug in CaesarCipher function. Check out my comment below.
> 
> * **[SUGGESTION]** Let's better organize your modules and tests and put them all inside the `src` folder, with subfolders for both your modules and another one for tests.
> * [x]  There are some missing test-cases in CaesarCipher:
>   
>   * It has a test for wrapping from `z` to `a`.
>   * It has a test for keeping the same case.
> 
> Please fix the below requirement and submit a new code review.
> 
> ##### Muhammad Ebeid

@mosaaleb Thank you for your suggestion. We rechecked the test with your input and its passing. Kindly re-check from your end.
testing

**Comment by mosaaleb** [comment]
Hi @imahnama Thanks for reaching out.

This is caused by a bug in your caesarCipher function (that's why we need to add more test cases), If you tried to change the order of these two test-cases, or write another test-case outside, it will fail:
Screenshot from 2020-08-25 22-18-48

The issue is that  `capsLetter` array doesn't get reset between function calls, It will have the same value from the past function call despite your are initializing it with `const capsLetters = [];`. This behavior is known as  [closures](https://javascript.info/closure). Please make sure to read about it.  Check out this answer to understand it more [closures](https://stackoverflow.com/a/19771024/4544219)

To solve the issue, you can reset `capsLetters` manually between function calls.

Feel free to contact me on Slack if you have further questions.

**Comment by OdongAlican** [comment]
hey @mosaaleb. Thanks a lot for the Feedback. We have implemented the requested changes. Could you please review the project and give us feedback.

**Review by Ispirett — APPROVED** [review]
## Status Approved :+1: 
Really good job @folks you did great work with all your changes made. :+1:

**Pull request merged**